subprojects/docs/src/docs/userguide/taskConfiguration.adoc

```diff
@@ -0,0 +1,269 @@
+// Copyright 2018 the original author or authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[avoiding_task_configuration]]
+== Avoiding Task Configuration
+
+We are asking early adopters to try out a new API for creating and configuring tasks to iron out any issues and gather feedback. This chapter provides a quick introduction to the feature and explains some guidelines for migrating your build to use the new API. We intend for this new API to replace the existing API through our usual deprecation process over several releases.
+
+[NOTE]
+====
+The new Task configuration API is currently <<feature_lifecycle,incubating>>. Please be aware that the DSL or API may change in minor Gradle versions. 
+====
+
+[[sec:what_is_new_task_configuration]]
+=== What is the new API good for?
+
+In a nutshell, the new API allows builds to avoid the cost of creating and configuring tasks during Gradle's configuration phase. Instead of immediately creating a task and running all configuration actions, Gradle exposes a api:org.gradle.api.provider.Provider[] to a task. We call these "lazy tasks". A task will only be created and configured when the task is required (usually execution). We identified task configuration as a contributor to overall configuration time, especially in large multi-project builds.
```

Comment:
I think we should emphasise a couple of things more strongly:

- Configuration time affects every build, and making it faster is good for everyone. Avoiding creating and configuring tasks that aren't used during the build (for example, don't create all the code quality, testing and publishing tasks if you're just building a jar) is one good way to do this.
- Sometimes configuration for certain tasks is expensive, and this can be ok if those tasks aren't used often by developers. This API allows you to implement this easily.

Reply:
We might also mention here, too, that your mileage may vary.  The most value comes from avoiding expensive task configuration, so a build that does not do much expensive configuration may not see as much improvement as one that does.  However, its still a best practice to defer _any_ configuration, so we recommend using this api wherever possible.